### PR TITLE
Adding support for loading VM modules from dynamic libraries.

### PR DESCRIPTION
**Pull request by benvanik**
Users can now produce native shared libraries that export one or more symbols used to create VM modules allowing for multiple different modules to be packaged in the same binary. These dynamic modules receive a key-value list of parameters that they can use to customize their behavior that can be programmatically specified by whatever is loading them (python kwargs, or flag `?key=value&key=value` on the command line).

On the hosting runtime side `iree_vm_dynamic_module_load_from_file` makes it easy to load and use the modules as if they were local to the binary. The command line tools have been updated to support multiple `--module=` flags and load each module into the context. This also allows for multiple vmfbs to be loaded and all kinds of linkage behavior (could have a training driver in a native module that calls functions in a compiled module loaded earlier, etc).

As part of this multiple `--module=` flags can be provided now to tools for both bytecode and dynamic modules. To ensure all the tools are using the same flag helper/loader I've switched iree-benchmark-module, iree-check-module, and iree-run-module to require explicitly specifying `--module=-` to load the module from stdin. This removes the need for the "reading from stdin message" that uglied up stdout.

I wasn't able to figure out Tracy support here (at least on Windows) as if tracing is globally enabled the shared libraries will get Tracy linked in and start a new tracing context. It's noted today and a warning is emitted when the sample is compiled with tracing enabled.

### CODE REVIEW

File: runtime/src/iree/tooling/context_util.c [hunk 1]
```diff
@@ -29,36 +31,29 @@
...
-// TODO(benvanik): module repeated flag. We could then allow the user to specify
-// either files or builtin module names in order to customize things. When we
-// support multiple types of dynamically loadable modules (lua/etc) we could
-// also allow mixes and use file ID snooping to choose a loader.
-IREE_FLAG(string, module, "-",
-          "File containing the module to load. Defaults to stdin (`-`).");
+IREE_FLAG_LIST(
+    string, module,
+    "A VM module to load; either a vmfb containing a compiled bytecode module\n"
+    "or a native system library containing a dynamic native module.");
```
Comment (ScottTodd) [inline runtime/src/iree/tooling/context_util.c]: Could mention if order matters here

File: runtime/src/iree/vm/dynamic/api.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// This should be implemented as pure as possible and may be called many times
+// while in process. |allocator| must be used for all allocations.
+//
+// The provided |max_version| is the maximum version the caller supports;
+// callees must return NULL if their lowest available version is greater
+// than the max version supported by the caller.
+typedef iree_status_t(IREE_API_PTR* iree_vm_dynamic_module_create_fn_t)(
+    iree_vm_dynamic_module_version_t max_version, iree_vm_instance_t* instance,
+    iree_host_size_t param_count, const iree_string_pair_t* params,
+    iree_allocator_t allocator, iree_vm_module_t** out_module);
```
Comment (ScottTodd) [inline runtime/src/iree/vm/dynamic/api.h]: "callees must return NULL" the function returns `iree_status_t` though?

File: samples/custom_module/dynamic/README.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
+# Dynamically loaded custom module sample
+
+DO NOT SUBMIT needs update
```
Comment (ScottTodd) [inline samples/custom_module/dynamic/README.md]: Reminder to resolve this before merge

File: samples/custom_module/dynamic/module.cc [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  IREE_RETURN_IF_ERROR(iree_custom_module_basic_register_types(instance));
+
+  // Create the custom module and return it to the runtime.
+  // NOTE: this isn't using the allocator here and that's bad.
+  auto module = std::make_unique<CustomModule>(
+      "custom", /*version=*/0, instance, host_allocator,
+      iree::span<const vm::NativeFunction<CustomModuleState>>(
+          kCustomModuleFunctions));
+  *out_module = module.release()->interface();
+  return iree_ok_status();
```
Comment (ScottTodd) [inline samples/custom_module/dynamic/module.cc]: Could this use the allocator? The NOTE is ominous (could reword to say why it's bad in general / why it's okay for this style of C++ application)

File: samples/custom_module/dynamic/README.md (hunk 3 above)
Comment (ScottTodd) [reply samples/custom_module/dynamic/README.md]: Bump - still has "DO NOT SUBMIT" text

Comment (benvanik) [reply samples/custom_module/dynamic/README.md]: 🤦

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**